tvb_contrib/tvb/contrib/scripts/plot/power_spectra_interactive.py

```diff
@@ -0,0 +1,634 @@
+# -*- coding: utf-8 -*-
+#
+#
+#  TheVirtualBrain-Scientific Package. This package holds all simulators, and 
+# analysers necessary to run brain-simulations. You can use it stand alone or
+# in conjunction with TheVirtualBrain-Framework Package. See content of the
+# documentation-folder for more details. See also http://www.thevirtualbrain.org
+#
+# (c) 2012-2022, Baycrest Centre for Geriatric Care ("Baycrest") and others
+#
+# This program is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software Foundation,
+# either version 3 of the License, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+# You should have received a copy of the GNU General Public License along with this
+# program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+#   CITATION:
+# When using The Virtual Brain for scientific publications, please cite it as follows:
+#
+#   Paula Sanz Leon, Stuart A. Knock, M. Marmaduke Woodman, Lia Domide,
+#   Jochen Mersmann, Anthony R. McIntosh, Viktor Jirsa (2013)
+#       The Virtual Brain: a simulator of primate brain network dynamics.
+#   Frontiers in Neuroinformatics (7:10. doi: 10.3389/fninf.2013.00010)
+#
+#
+
+"""
+An interactive power spectra plot generated from a TVB TimeSeries datatype.
+
+Usage
+::
+
+    #Load the demo data
+    import numpy
+    data = numpy.load("demos/demo_data_region_16s_2048Hz.npy")
+    period = 0.00048828125 #NOTE: Providing period in seconds
+
+    #Create a tvb TimeSeries object
+    import tvb.datatypes.time_series
+    tsr = tvb.datatypes.time_series.TimeSeriesRegion()
+    tsr.data = data
+    tsr.sample_period = period
+    tsr.sample_period_unit = 's'
+
+    #Create and launch the interactive visualiser
+    import tvb.simulator.power_spectra_interactive as ps_int
+    psi = ps_int.PowerSpectraInteractive()
+    psi.time_series = tsr
+    psi.show()
+
+
+.. moduleauthor:: Stuart A. Knock <Stuart@tvb.invalid>
+
+"""
+
+import numpy
+import pylab
+import matplotlib.widgets as widgets
+from tvb.simulator.common import get_logger
+import tvb.datatypes.time_series as time_series_datatypes
+from tvb.basic.neotraits.api import HasTraits, Attr, Int
+
+
+LOG = get_logger(__name__)
+
+# Define a colour theme... see: matplotlib.colors.cnames.keys()
+BACKGROUNDCOLOUR = "slategrey"
+EDGECOLOUR = "darkslateblue"
+AXCOLOUR = "steelblue"
+BUTTONCOLOUR = "steelblue"
+HOVERCOLOUR = "blue"
+
+
+class PowerSpectraInteractive(HasTraits):
```

Comment:
Why do you need to copy this class in contrib?
Can't you at least inherit from the similarly named class from under `tvb.simulator.plot`?

Reply:
Let me try! I remember that there were smaller or bigger changes all over that class. I also found some bugs there. But, I need to look at it again more carefully. I will come back to you about this.

Reply:
we would happily accept bug fixes in `tvb.simulator.plot` itself.
I think this widget was build mostly for your usages, thus no harm to change it as you like, as long as it still works.

Reply:
> we would happily accept bug fixes in `tvb.simulator.plot` itself. I think this widget was build mostly for your usages, thus no harm to change it as you like, as long as it still works.

So, I have inherited it from tvb.simulator.plot and also renamed it as my widget plots also coherence.

Reply:
Cool about the inherit. I still see some place to improve, for example: can't the `PowerSpectraCoherenceInteractive.__init__` and `PowerSpectraCoherenceInteractive.create_figure` methods be smaller (by removing the duplicate code which already exists in the parent)?
Also, sonar complains on 14 code smells. Some are quite easy to fix before we merge (e.g. remove commented code)

Reply:
> Cool about the inherit. I still see some place to improve, for example: can't the `PowerSpectraCoherenceInteractive.__init__` and `PowerSpectraCoherenceInteractive.create_figure` methods be smaller (by removing the duplicate code which already exists in the parent)? Also, sonar complains on 14 code smells. Some are quite easy to fix before we merge (e.g. remove commented code)

Yes, I could reduce the duplicate code, but only if I refactor the original tvb.simulator.plot class.

So, let me try. I will also, then, check the smelly code...!